File: usaco_training/4/buylow.java
/*
ID: rednano1
PROG: buylow
LANG: JAVA
*/
import java.io.*;
import java.util.*;
import java.math.BigInteger;

public class buylow {
    public static void main(String[] args) throws IOException {
        final Scanner scan = new Scanner(new File("buylow.in"));
        final PrintWriter pw = new PrintWriter(new File("buylow.out"));

        final int n = scan.nextInt();
        final int[] a = new int[n];
        for (int i = 0; i < n; i++) {
            a[i] = scan.nextInt();
        }
        scan.close();

        int bestLength = 1;
        final int[] dp1 = new int[n];
        final BigInteger[] dp2 = new BigInteger[n];

        for (int i = 0; i < n; i++) {
            int length = 1;
            dp2[i] = BigInteger.ONE;
            BigInteger count = BigInteger.ONE;
            final Set<Integer> previous = new HashSet<Integer>();

            for (int j = i - 1; j >= 0; --j) {
                if (a[i] < a[j] && length < dp1[j] + 1) {
                    length = dp1[j] + 1;
                    count = dp2[j];
                    previous.clear();
                    previous.add(a[j]);
                } else if (a[i] < a[j] && length == dp1[j] + 1 && !previous.contains(a[j])) {
                    count = count.add(dp2[j]);
                    previous.add(a[j]);
                }
            }

            dp1[i] = length;
            dp2[i] = count;
            bestLength = Math.max(bestLength, length);
        }

        BigInteger totalCount = BigInteger.ZERO;
        final Set<Integer> last = new HashSet<Integer>();
        for (int i = n - 1; i >= 0; i--) {
            if (dp1[i] == bestLength && !last.contains(a[i])) {
                totalCount = totalCount.add(dp2[i]);
                last.add(a[i]);
            }
        }

        pw.println(bestLength + " " + totalCount);
        pw.close();

        /*
        for (int i = 0; i < n; i++) {
            System.out.print(a[i] + " ");
        }
        System.out.println();

        for (int i = 0; i < n; i++) {
            System.out.print(dp1[i] + " ");
        }
        System.out.println();

        for (int i = 0; i < n; i++) {
            System.out.print(dp2[i] + " ");
        }
        System.out.println();
        */
    }
}

File: usaco_training/4/buylow.cpp
/*
ID: rednano1
PROG: buylow
LANG: C++11
*/
#include <iostream>
#include <fstream>
#include <cstring>
#include <algorithm>
#include <set>
#include <utility>

using namespace std;

int n;
int a[5003];
// length of the LDS ending here
int dp1[5003];
// Number of ways to get to an LDS of maximum length ending here.
int dp2[5003];

int main() {
    ifstream fin("buylow.in");
    ofstream fout("buylow.out");

    fin >> n;
    for (int i = 0; i < n; ++i) {
        fin >> a[i];
    }
    fin.close();

    int best_length = 1;
    memset(dp1, 0, sizeof(dp1));
    memset(dp2, 0, sizeof(dp2));

    for (int i = 0; i < n; ++i) {
        int length = 1;
        int count = 1;
        set<int> prev_val;
        for (int j = i - 1; j >= 0; --j) {
            if (a[i] < a[j] and length < dp1[j] + 1) {
                length = dp1[j] + 1;
                count = dp2[j];
                prev_val.clear();
                prev_val.insert(a[j]);
            } else if (a[i] < a[j] and length == dp1[j] + 1 and prev_val.find(a[j]) == prev_val.end()) {
                cout << i << " " << j << " " << count << " " <<  dp2[j] << '\n';
                count += dp2[j];
                prev_val.insert(a[j]);
            }

            cout << i << ' ' << j << ' ' << count << '\n';
        }

        dp1[i] = length;
        dp2[i] = count;
        best_length = max(best_length, length);
    }

    long long total_count = 0LL;
    set<int> last;
    for (int i = n - 1; i >= 0; --i) {
        if (dp1[i] == best_length && last.find(a[i]) == last.end()) {
            total_count += dp2[i];
            last.insert(a[i]);
        }
    }

    fout << best_length << ' ' << total_count << '\n';
    fout.close();

    for (int i = 0; i < n; ++i)
        cout << a[i] << ' ';
    cout << '\n';

    for (int i = 0; i < n; ++i)
        cout << dp1[i] << ' ';
    cout << '\n';

    for (int i = 0; i < n; ++i)
        cout << dp2[i] << ' ';
    cout << '\n';
    return 0;
}
